wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,709 @@
+# cuML Python Estimators Developer Guide
+
+This guide is meant to help developers follow the correct patterns when creating/modifying any class that inherits from `cuml.common.base.Base`.
+
+Start by reading the following:
+
+1. [SkLearn Estimator Docs](https://scikit-learn.org/stable/developers/develop.html)
+   1. `cuml` Estimator design follows Sklearn very closely. We will only cover portions where our design differs from this document
+   2. Pay close attention to these sections as these topics have caused pain points in the past:
+      1. [Instantiation](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+      2. [Estimated Attributes](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+      3. [`get_params` and `set_params`](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+      4. [cloning](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+2. [SkLearn Docstring Guide](https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation)
+   1. We follow the same guidlines for speficying array-like objects, array shapes, dtypes, and default values
+
+## Background
+
+Some background is necessary to understand the design of estimators and how to work around any non-standard situations.
+
+### Input and Output Types in cuML
+
+In `cuml` we support both ingesting and generating a variety of different n-dimensional array types. Estimators should be able to accept and return any array type. The types that are supported as of release 0.17:
+
+ - cuDF DataFrame or Series
+ - Pandas DataFrame or Series
+ - NumPy Arrays
+ - Numba Device Arrays
+ - CuPy arrays
+ - CumlArray type (Internal to the `cuml` API only.)
+
+When converting between types, it's important to minimize the CPU<->GPU type conversions as much as possible. Conversions such as NumPy -> CuPy or Numba -> Pandas DataFrame will incur a performance penalty as memory is copied from device to host or viceversa.
+
+Converting between types of the same device, i.e. CPU<->CPU or GPU<->, do not have as significant of a penalty.
+
+Finally, conversions between Numba<->CuPy<->CumlArray incur the least amount of overhead since only the device pointer is moved from one class to another.
+
+Internally, all array's should be converted to `CumlArray` as much as possible since it is compatible with all output types and can be easily converted.
+
+### Speficying the Array Output Type
+
+Users can choose which array type should be returned by cuml by either: 
+1. Individually setting the output_type property on an estimator class (i.e `Base(output_type="numpy")`)
+2. Globally setting the `cuml.global_output_type`
```

Comment:
I left off the `using_output_type` context manager because all it is doing is setting and restoring `cuml.global_output_type`. I just didnt think it was fundamentally a different way of controlling the output type.

I have no problems adding it in here

Reply:
Added in `using_output_type`. Resolving